### ISSUE TITLE
Resend doesn't send the last value

### ISSUE THREAD

**Issue by jeromeminardi** [bartbutenaers/node-red-contrib-msg-resend#5]
When sending multiple values to msg-resend in a short amount of time, the last value is not resent but the previous one. 

Please see the demonstration flow and the trace : 

5 Feb 14:05:10 - [info] [function:Debug] --------
5 Feb 14:05:10 - [info] [function:Debug] > NEW TIMESTAMP : 1517835910101
5 Feb 14:05:10 - [info] [function:Debug Resend] MSG RESEND : 1517835910101
5 Feb 14:05:15 - [info] [function:Debug Resend] MSG RESEND : 1517835910101
5 Feb 14:05:20 - [info] [function:Debug Resend] MSG RESEND : 1517835910101
5 Feb 14:05:22 - [info] [function:Debug] --------
5 Feb 14:05:22 - [info] [function:Debug] > NEW TIMESTAMP : 1517835922425
5 Feb 14:05:22 - [info] [function:Debug Resend] MSG RESEND : 1517835922425
5 Feb 14:05:27 - [info] [function:Debug Resend] MSG RESEND : 1517835922425
5 Feb 14:05:30 - [info] [function:Debug] --------
5 Feb 14:05:30 - [info] [function:Debug] > NEW TIMESTAMP : 1517835930980
5 Feb 14:05:30 - [info] [function:Debug Resend] MSG RESEND : 1517835930980
5 Feb 14:05:32 - [info] [function:Debug] --------
5 Feb 14:05:32 - [info] [function:Debug] > NEW TIMESTAMP : 1517835932040
5 Feb 14:05:32 - [info] [function:Debug Resend] MSG RESEND : 1517835932040
5 Feb 14:05:35 - [info] [function:Debug Resend] MSG RESEND : 1517835930980
5 Feb 14:05:40 - [info] [function:Debug Resend] MSG RESEND : 1517835930980

The first two blocks are correct, then I click two times on the trigger quickly, and the message resent by msg-resent is the previous one after a first correct output (1517835930980 instead of 1517835932040).

The flow for reproducing is the following : 

[{"id":"a0ca31a4.f98468","type":"tab","label":"Test MSG Resend","disabled":false,"info":""},{"id":"c9c56931.330ed","type":"inject","z":"a0ca31a4.f98468","name":"","topic":"test_resend","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"x":205.5,"y":286,"wires":[["3c6d9799.32f468","b74edcda.b17698"]]},{"id":"b74edcda.b17698","type":"msg-resend","z":"a0ca31a4.f98468","interval":"5","maximum":"-1","bytopic":false,"clone":false,"name":"","x":423.5,"y":208,"wires":[["792a11a5.a5c04"]]},{"id":"792a11a5.a5c04","type":"function","z":"a0ca31a4.f98468","name":"Debug Resend","func":"node.log(\"MSG RESEND : \" + msg.payload)\nreturn msg;","outputs":1,"noerr":0,"x":632.5,"y":208,"wires":[[]]},{"id":"3c6d9799.32f468","type":"function","z":"a0ca31a4.f98468","name":"Debug","func":"node.log(\"--------\")\nnode.log(\"> NEW TIMESTAMP : \" + msg.payload)\nreturn msg;","outputs":1,"noerr":0,"x":404,"y":367,"wires":[[]]}]

**Comment by bartbutenaers** [comment]
Hi Djé,

As you can see in the readme file, this node does [message frame limitation](https://github.com/bartbutenaers/node-red-contrib-msg-resend#message-rate-limitation): _when messages arrive to quickly, it is useless to create continiously new timers (which don't get the time to respond)._

But indeed there seems to be a bug in this mechanism:  
![image](https://user-images.githubusercontent.com/14224149/35829870-97e142ae-0ac4-11e8-83e7-0af3ef0ccb05.png)

While 1517835930980 is being resended, the new message 1517835932040 will be **discarded** since it arrives too quickly:
1. The message (1517835932040 ) will be send to the output for the first time
2. But since it arrives to early, it (1517835932040 ) will be discarded. 

However this way 1517835932040 will be sended once, which was not my intention: message 1517835932040  should **never** be sended.

**FIX:** When I change the code to execute first (2) and afterwards (1), the problem seems to be solved.  

However it might be useful perhaps if this mechanism could be disabled, by adding a checkbox to the config screen:
![image](https://user-images.githubusercontent.com/14224149/35831196-364e9b04-0ac9-11e8-82c9-b63bcb070b09.png)

I introduced the mechanism because if you keep sending messages at high speed, the timers won't get a chance to start and do their job.  But if you only have this situation from time to time, it might be a requirement that always the last message is being resend (even if it arrives too quickly).

As you can see in the latter screenshot, there are a series of other new functionalities already available for a new version of the resend node.  I will test the new version as soon as possible, so you can check whether it fits your needs.

P.S. Thanks for the nice reproduction flow !

Kind regards,
Bart

**Comment by bartbutenaers** [comment]
Hi Djé,

I had yesterday problems, when I tried to import your test flow.  Thought it was a general Node-Red problem, so I started a new [discussion](https://groups.google.com/forum/#!topic/node-red/r4LZs8c9nc0) on the Node-Red forum.  Could you please let me know how you exported your flow, so I can understand why the flow became unimportable.

Thanks !
Bart

**Comment by jeromeminardi** [comment]
Hi Bart, 

Ok I found the issue, it's because of the copy paste in this github without specifying github it's code.
Here is the correct flow using the "code" feature.

`[{"id":"c9c56931.330ed","type":"inject","z":"a0ca31a4.f98468","name":"","topic":"test_resend","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"x":205.5,"y":286,"wires":[["3c6d9799.32f468","b74edcda.b17698"]]},{"id":"b74edcda.b17698","type":"msg-resend","z":"a0ca31a4.f98468","interval":"5","maximum":"-1","bytopic":false,"clone":false,"name":"","x":423.5,"y":208,"wires":[["792a11a5.a5c04"]]},{"id":"792a11a5.a5c04","type":"function","z":"a0ca31a4.f98468","name":"Debug Resend","func":"node.log(\"MSG RESEND : \" + msg.payload)\nreturn msg;","outputs":1,"noerr":0,"x":632.5,"y":208,"wires":[[]]},{"id":"3c6d9799.32f468","type":"function","z":"a0ca31a4.f98468","name":"Debug","func":"node.log(\"--------\")\nnode.log(\"> NEW TIMESTAMP : \" + msg.payload)\nreturn msg;","outputs":1,"noerr":0,"x":404,"y":367,"wires":[[]]}]`

**Comment by jeromeminardi** [comment]
_I added another comment but don't see it so I retype it._ 

First, thank a lot for your quick support and a solution, that's cool :)

Secondly, I can propose you an alternative solution a little bit more complex but with advantages (Hope it's doable within NodeRed programming model, I'm not an expert of NodeRed).

Instead of starting a new timer each time you receive a message, I propose you to decouple timers and messages reception. What you can do is the following algorithm:

```
If new_message received
    If timer not exist then create timer endif
    message_to_publish = new message
endif

Callback : 
When timer need to send the message to repeat
    send message_to_publish

```
With this, each time a message is received, if the timer exist then it just overwrite the previous message by the new one whatever the received frequency. Of course it can be adapted to multi topics with a dictionary for storing last messages for each topic.

Hope I'm clear, don't hesitate to ask me question if I'm not.

Regards,
Djé.

**Comment by bartbutenaers** [comment]
Hi Jerome,

Well that is indeed not a bad idea, to let the current timer send the new input value.  Have not tested it yet, but I don't see why it shouldn't work.  A new timer has only to be created when new settings are applied (e.g. new interval length specified via an input message).   Reusing timers could result in quite better performance.   Never thought about it at the time being, because this 'resend' node was my first Javascript program ...

However I'm afraid that I could damage some existing flows, by introducing this new mechanism.  Indeed **currently** my node works like this:

![image](https://user-images.githubusercontent.com/14224149/35935791-c4ea2dc8-0c41-11e8-9ae0-669d7cf091f2.png)

As soon as a new message msg2 arrives, the resending of msg1 will be stopped.  Msg2 will get a new timer...

When I would use **your mechanism** (i.e. reuse the current timer of msg1 to resend msg2), the output would become quite different:

![image](https://user-images.githubusercontent.com/14224149/35935958-40afd34a-0c42-11e8-948e-8617c57bf82a.png)

Indeed the timer keeps running at it's own pace: instead of the last msg1, now the msg2 will be send (again) ...

This might be a problem for some users.  I simply don't know for what use cases my contribution is being used all over the world.  I could add an extra textbox to the config screen 'reuse timers (for performance)', but don't know if my node would become incomprehensible.

Any thoughts are welcome, before I start coding on this...

Kind regards,
Bart

**Comment by jeromeminardi** [comment]
Hum... I understand your point, the behavior is indeed different so it's an issue because, I agree, it impact all potential users. I'm not sure it worth adding a second behaviour in the pluging. To be honest, I think that simple is beautiful so having two behavior in one plugin is not the right thing to do. 

I can change my code for adapting to the current behavior, so I would say that the first correction (for avoiding 1517835932040 to be sent) will do the job. I will think about it, but I propose you to close the case with this simple fix for now if you agree. Are you ok with that ?

I would like to thank you a lot for your support. I really appreciated the good discussion. Djé.

**Comment by bartbutenaers** [comment]
Deal ;-)

It was indeed a good discussion and a good proposal.  Pitty that I didn't think about it myself, when I created this node...  But anyway I will create in the next week a 0.0.2 version, with your fix and all kind of other enhancements.  Will keep you informed.  Bart

**Comment by bartbutenaers** [comment]
He Djé,

I have put version **0.0.7** on Github.  It contains your fix, but also lots of extra new stuff (as you can see in the readme file).  That version is not published on NPM yet, so you can install it directly from Github like this:

`npm install bartbutenaers/node-red-contrib-msg-resend`

Please let me know if it fits your needs.  Thanks !

Bart

**Comment by jeromeminardi** [comment]
Hi Bart, I've tested the new version and it works pretty well. Thanks for this fix, now it work exactly how I was expecting with the high rate option.
Thanks and Take care, Djé.

**Comment by bartbutenaers** [comment]
Hey Djé,

Thanks for keeping me updated!

I have published version **0.0.7** on NPM.  So from here on you can stall it again using the normal npm commands.  Great doing business with you ;-)

See you around,
Bart
